### PR TITLE
Basic WebGL Backend

### PR DESCRIPTION
**Pull request by phisiart**
### TLDR

Currently the following demo program runs and gets the correct result.

```python
from __future__ import absolute_import, print_function

import tvm
import numpy as np

n = tvm.var("n")
A = tvm.placeholder((n,), name='A')
B = tvm.placeholder((n,), name='B')
C = tvm.compute(A.shape, lambda i: A[i] + B[i], name="C")

s = tvm.create_schedule(C.op)
s[C].opengl()

fadd_gl = tvm.build(s, [A, B, C], "opengl", name="myadd")
print("------opengl code------")
print(fadd_gl.imported_modules[0].get_source(fmt="gl"))

ctx = tvm.opengl(0)
n = 10
a = tvm.nd.array(np.random.uniform(size=(n)).astype(A.dtype), ctx)
b = tvm.nd.array(np.random.uniform(size=(n)).astype(B.dtype), ctx)
c = tvm.nd.array(np.zeros((n), dtype=C.dtype), ctx)
fadd_gl(a, b, c)

np.testing.assert_allclose(c.asnumpy(), a.asnumpy() + b.asnumpy())
```

The corresponding fragment shader is

```glsl
#version 330 core
uniform sampler2D A;
uniform sampler2D B;
out float C;
void main() {
  ivec2 threadIdx = ivec2(gl_FragCoord.xy);
  C = (texelFetch(A, ivec2(threadIdx.x, 0), 0).r + texelFetch(B, ivec2(threadIdx.x, 0), 0).r);
}
```

### OpenGL Tensor Storage
- We store tensors in OpenGL textures.
- No matter what dimensions a tensor has, we always store it in a 2D texture with height=1.
- The reason we are not using 1D textures is that `texelFetch` in GLSL only supports 2D textures.
- We support uint{8,16,32}, int{8,16,32} and float32 textures.

### OpenGL Schedule
- We added an `opengl` schedule, which basically fuses all dimensions into one and binds that single dimension to threadIdx.x.

### Codegen
- We haven't changed lowering at all. When the IR says `Store(buffer, index)` and the buffer happens to be the output texture, we check that index must be `threadIdx.x`, and emit code to output a pixel.
- The codegen part has only been started. We will go through all the AST nodes as our next step.

### Task Items
- <del>Remove the dependency on glfw and glad.</del> Emscripten supports glfw. Glad removed.
- <del>Investigate emscripten.</del> Runtime runnable.
- <del>Support other types than float.</del> Done.
- <del>Use full RGBA channels.</del> Not critical now.
- <del>Don't let height always be 1. The reason is that OpenGL textures have stupid size limitations. For example, you can have a 2500x2500 2D texture but not a 6250000x1 2D texture. One possible way is just let height be the maximum supported value.</del> Not critical now.
- Go through the AST to properly do codegen.
- <del>Don't fuse reduction.</del> Gemm runnable.
- <del>Fix argument name generation.</del> Done.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -7,3 +7,7 @@
 [submodule "dlpack"]
 	path = dlpack
 	url = https://github.com/dmlc/dlpack
+[submodule "glad"]
```
Comment (tqchen) [inline .gitmodules]: What is glad used for, is it possible to only use OpenGL standard and not rely on new libraries?

File: include/tvm/runtime/c_runtime_api.h [hunk 2]
```diff
@@ -57,8 +57,9 @@ typedef int64_t tvm_index_t;
 typedef enum {
   // Extension DRAM type, used for quickly test extension device
   // The device api can differ depending on the xpu driver registered.
-  kExtDev = 12
+  kExtDev = 12,
   // AddExtraTVMType which is not in DLPack here
+  kOpenGL = 11,
```
Comment (tqchen) [inline include/tvm/runtime/c_runtime_api.h]: put kOpenGL before kExtDev

File: .gitmodules (hunk 1 above)
Comment (phisiart) [reply .gitmodules]: We actually rely on 2 libraries.

- `glad`: This is the library to help you use the correct version of OpenGL.
  For example, suppose your OS supports OpenGL 3.3, then the OS is able to give you any OpenGL version below 3.3 on your request. However, if you just `#include <GL/gl.h>` or whatever header your OS provides, the function prototypes are for OpenGL 1.3 or something, so you can't use OpenGL 3.3 even your OS supports it. The OS provides a way of querying the function pointers of OpenGL 3.3 APIs. That's what `glad` does. It is possible that we don't rely on `glad`, but with a huge cost - we must write our own code to get all the function pointers.

- glfw: This is the library to help you create an OpenGL context in a cross-platform way.
  You can't use any OpenGL API before creating a context. GLFW wraps the context creation code of different OS'es to provide a unified API. It is possible that we don't rely on `glfw`, but with a huge cost - we must write context initialization code for different platforms.

I can remove the dependencies on those libraries, but that can only happen after Dec 12 or something. We are focusing on producing a final-report-able version currently.

Comment (tqchen) [reply .gitmodules]: Sure, I think these are fine as long as we fix this at the final merge.

File: src/schedule/schedule_lang.cc [hunk 3]
```diff
@@ -397,6 +397,25 @@ Stage& Stage::double_buffer() {
   return *this;
 }
 
+Stage& Stage::opengl() {
+  CHECK(!is_scheduled()) << "Must be a fresh schedule";
+  StageNode *self = operator->();
+
+  auto all_iter_vars = self->all_iter_vars; // curr version of all_iter_vars
```
Comment (tqchen) [inline src/schedule/schedule_lang.cc]: be careful that this can include reduction variables, which should not be part of the fusing stage

File: src/runtime/opengl/opengl_device_api.cc [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+  glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+  glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
+  window_ = glfwCreateWindow(kWindowWidth, kWindowHeight, "", nullptr, nullptr);
+  if (window_ == nullptr) {
+    LOG_ERROR.stream() << "glfwCreateWindow() failed!";
+    assert(false);
+  }
+
+  LOG_INFO.stream() << "GLFW says OpenGL version: "
```
Comment (tqchen) [inline src/runtime/opengl/opengl_device_api.cc]: Can directly use LOG(INFO)

File: src/codegen/codegen_opengl.cc [hunk 5]
```diff
@@ -0,0 +1,138 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file codegen_opengl.cc
```
Comment (tqchen) [inline src/codegen/codegen_opengl.cc]: Add a specific comment that we are targeting subset of OpenGL that is working for WebGL2,(no compute shaders)

File: python/tvm/target.py [hunk 6]
```diff
@@ -114,6 +114,9 @@ def __init__(self,
         elif target_name in ("metal",):
             self.keys += ("gpu",)
             self.max_num_threads = 256
+        elif target_name in ("opengl"):
```
Comment (tqchen) [inline python/tvm/target.py]: use opengl as key, as opengl schedule need to be different from gpu schedules

Comment (PENGUINLIONG) [inline python/tvm/target.py]: Shouldn't it be `("opengl",)`? It ought to be searching for `target name` in a tuple.

**Comment by phisiart** [comment]
I'm currently stuck on other final projects for courses, so I won't be able to work on this PR until mid December. Then I should have roughly a month after the end of this semester.

File: python/tvm/target.py (hunk 6 above)
Comment (phisiart) [reply python/tvm/target.py]: Fixed.

File: src/codegen/codegen_opengl.cc (hunk 5 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: Added.

File: src/schedule/schedule_lang.cc (hunk 3 above)
Comment (phisiart) [reply src/schedule/schedule_lang.cc]: Will go back to this later. Added a TODO for now.

File: include/tvm/runtime/c_runtime_api.h (hunk 2 above)
Comment (phisiart) [reply include/tvm/runtime/c_runtime_api.h]: Moved.

**Comment by phisiart** [comment]
Next step: Remove dependency on glfw and glad. I will focus on Ubuntu, since it should be the tricky one. Windows and Mac should have official APIs to create OpenGL contexts.

I will see what Ubuntu Desktop has by default and only depend on that.
On Ubuntu Server, OpenGL might not exist by default (because it doesn't have GUI). If that's the case I will see what minimum packages are needed to have OpenGL, and only depend on them.

**Comment by PENGUINLIONG** [comment]
Use `wgl*` functions you can create a context on Windows.

On Linux, if MESA is installed, OpenGL soft-rendering should be available as well as accelerated graphics, if there is a GPU. Compute Shader should be supported as it's a part of the OpenGL 4.3 spec.

FYI, without GLFW it will be your responsibility to query if a GL function is available. Using GL_ABR_compute_shader, the minimum GL version requirement should be (Core) 4.3. On initialization, GLFW stores function pointers to GL APIs so that the query results can be reused.

File: .gitmodules (hunk 1 above)
Comment (PENGUINLIONG) [reply .gitmodules]: AFAIK GLFW itself wraps up all the functions. Maybe GLAD is not needed since GLFW is already there. If my memory went wrong please inform me.

**Comment by tqchen** [comment]
One of the the main goal here is to use WebGL, this means

- so the standard version should be less than OpenGL3.
- we cannot use compute shader

Please try to emscripten the GL runtime to see if we can successfully build it

**Comment by PENGUINLIONG** [comment]
I see. But it would be nice if we can utilize the functionalities provided by OpenGL of higher versions - Compute Shader. We have Compute Shader support in OpenGL (Core) 4.3 and in OpenGL ES 3.1 but not yet in WebGL. In the current latest commit, you can see there is a [dummy vertex shader](https://github.com/phisiart/tvm/blob/e58817058fa1cac783556e8befd31be872851964/src/runtime/opengl/opengl_device_api.cc#L318-L325) that doesn't need to but has to be executed.

Or a strategy can be devised so that both shaders can do some work.

**Comment by tqchen** [comment]
For most other devices, opencl is also available, which is more desirable than OpenGL itself

**Comment by phisiart** [comment]
After some experiments, I found that the following is the minimum requirements for `find_package(OpenGL REQUIRED)` in cmake to succeed.

Dockerfile:
```
FROM ubuntu

RUN apt-get update --fix-missing

RUN apt-get install -y --no-install-recommends \
  make cmake g++ libgl1-mesa-dev
```

CMake is looking for `libGL.so`, and on Ubuntu there are 3 packages that provide this (see [here](https://packages.ubuntu.com/search?suite=precise&arch=amd64&mode=filename&searchon=contents&keywords=libGL.so)):

- mesa: This is what comes by default when you install Ubuntu desktop;
- nvidia: The NVIDIA driver;
- fglrx: The AMD driver.

Therefore, we should depend on `libgl1-mesa-dev`.

**Comment by phisiart** [comment]
Okay, I just tested and found out that on a fresh installation of Ubuntu Desktop you also need the package `libgl1-mesa-dev`, for `find_package(OpenGL REQUIRED)` in cmake to succeed (because Ubuntu doesn't come with the 'dev' version of the package).

**Comment by PENGUINLIONG** [comment]
Since the goal for the OpenGL port has shifted to NN deployment on web apps.. Does it still necessarily need mesa?

Emscripten is compiling LLVM intermediate to JavaScript. It should require no linking to native libraries. Maybe we can just `extern "C"`-declare those APIs we need, so that the feature can be implemented with no dependency added.

See [this](https://www.khronos.org/registry/OpenGL/api/GLES2/gl2.h) for GLES2 APIs.

**Comment by phisiart** [comment]
@PENGUINLIONG I think we want to support both running natively and in the browser.

A funny thing is that emscripten has direct support for glfw. If we remove the dependency on glfw then we would be calling native glx functions. I'm not sure whether emscripten has good support for those. I will look into it and report my findings.

**Comment by PENGUINLIONG** [comment]
@phisiart It seems Emscripten simply map the calls. See [this](https://github.com/kripken/emscripten/blob/master/src/library_gl.js). Then it's possible that we simply declare them and they will work.

However, it seems not possible we can create context without dependencies.

**Comment by tqchen** [comment]
I think the first milestone is to confirm emscripten with simple gl runtime works. You can likely utilize the RPC module here https://github.com/dmlc/tvm/tree/master/web for testing

**Comment by phisiart** [comment]
I tried RPC. I have successfully crashed in the middle of OpenGL initialization, after `glfwInit()`. It seems like incompatibility with glad. Since I'm going to remove the dependency on glad anyway, let me do it and then continue with RPC.

**Comment by phisiart** [comment]
Update: Now OpenGL initialization succeeds.

**Comment by phisiart** [comment]
Update: Now rendering succeeds.

**Comment by tqchen** [comment]
here are some trackable milestones changes that I think could be useful. It would be very helpful to create a series of test functions under tests/web/webgl that relies on web proxy rpc and can be run manually.


- [ ] Runtime array copy pass, something like https://github.com/dmlc/tvm/blob/master/tests/python/unittest/test_runtime_ndarray.py
- [ ] Simple ewise pass via RPC, something like https://github.com/dmlc/tvm/blob/master/tests/python/unittest/test_runtime_rpc.py#L134
- [ ] Pass the basic gemm
- [ ] Enable gl in topi, so we can pass end to end resnet in nnvm compiler

File: Makefile [hunk 7]
```diff
@@ -30,10 +30,10 @@ CFLAGS = -std=c++11 -Wall -O2 $(INCLUDE_FLAGS) -fPIC
 FRAMEWORKS =
 OBJCFLAGS = -fno-objc-arc
 EMCC_FLAGS= -std=c++11 -DDMLC_LOG_STACK_TRACE=0\
-	-Oz -s RESERVED_FUNCTION_POINTERS=2 -s MAIN_MODULE=1 -s NO_EXIT_RUNTIME=1\
+	-O0 -s RESERVED_FUNCTION_POINTERS=2 -s MAIN_MODULE=1 -s NO_EXIT_RUNTIME=1\
```
Comment (tqchen) [inline Makefile]: is -O0 necessary? or can we use -Oz?

File: include/tvm/runtime/device_api.h [hunk 8]
```diff
@@ -55,11 +55,13 @@ class DeviceAPI {
...
    * \brief Allocate a data space on device.
    * \param ctx The device context to perform operation.
-   * \param size The size of the memory
+   * \param type The type of elements.
+   * \param nbytes The number of bytes in memory
    * \param alignment The alignment of the memory.
    * \return The allocated device pointer
    */
-  virtual void* AllocDataSpace(TVMContext ctx, size_t size, size_t alignment) = 0;
+  virtual void* AllocDataSpace(TVMContext ctx, TVMType type, size_t nbytes,
```
Comment (tqchen) [inline include/tvm/runtime/device_api.h]: let us put type->type_hint and put it as last parameter. Use comment to say type_hint is only needed by a few backend such as GL

File: src/runtime/cpu_device_api.cc [hunk 9]
```diff
@@ -20,13 +20,14 @@ class CPUDeviceAPI final : public DeviceAPI {
       *rv = 1;
     }
   }
-  void* AllocDataSpace(TVMContext ctx, size_t size, size_t alignment) final {
+  void* AllocDataSpace(TVMContext ctx, TVMType type, size_t nbytes,
```
Comment (tqchen) [inline src/runtime/cpu_device_api.cc]: if we have multiple line breaks, it might be more natural to break each argument into one line

File: src/runtime/rpc/rpc_session.h [hunk 10]
```diff
@@ -48,6 +48,7 @@ enum class RPCCode : int {
   kModuleFree,
   kModuleGetFunc,
   kModuleGetSource,
+  kTestRemoteOpenGL,
```
Comment (tqchen) [inline src/runtime/rpc/rpc_session.h]: this can be safely removed when we upstream?

File: include/tvm/schedule.h [hunk 11]
```diff
@@ -213,6 +213,11 @@ class Stage : public NodeRef {
    * \return reference to self.
    */
   Stage& double_buffer();   // NOLINT(*)
+  /*!
+   * \brief Schedule for GpenGL fragment shader.
```
Comment (tqchen) [inline include/tvm/schedule.h]: OpenGL

**Comment by tqchen** [comment]
for the case when we need 2D texture, we might be able to use the following workaround.

Always allocate memory as i = 2^k * y + x, where x and y are two dimensional axis. k is the folding ideal packing length over the x dimension, say 10

so we can easily use bit operation to get y and x from global index i, and reconstruct vice versa

File: Makefile (hunk 7 above)
Comment (phisiart) [reply Makefile]: Oh this is just for debugging purposes. Will be changed back when I cleanup.

File: src/runtime/rpc/rpc_session.h (hunk 10 above)
Comment (phisiart) [reply src/runtime/rpc/rpc_session.h]: Yes, this is just for my debugging purposes. Will remove when I cleanup.

**Comment by phisiart** [comment]
Update: I'm working on webgl + rpc. I'm trying to upload and download an OpenGL module. I've successfully crashed in the middle of (re-)creating the shader in the browser.

**Comment by phisiart** [comment]
Update: Sorry for the delay. I can save and load OpenGL module locally, but not remotely.

I added my test scripts (not final version):
`tests/webgl/test_local_save_load.py` works
`tests/webgl/test_remote_save_load.py` crashes when running the remote module `fhost(a, b, c)`.

I think I'm doing something wrong. But I haven't had the time to dive into the RPC part.

**Comment by phisiart** [comment]
Update: Now the entire flow completes! I did add some hack, and the result array is incorrect, but that's problem inside the webgl part. I'll fix that tomorrow.

File: src/runtime/opengl/opengl_module.cc [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+void OpenGLWrappedFunc::operator()(TVMArgs args, TVMRetValue* rv,
+                                   void** void_args) const {
+  LOG(INFO) << "OpenGLWrappedFunc::operator()";
+
+  gl::Texture* arg0 = *static_cast<gl::Texture**>(void_args[0]);
+  std::vector<std::pair<std::string, gl::Texture*>> name_texture_pairs;
+  for (size_t i = 1; i != arg_size_.size(); ++i) {
+    name_texture_pairs.push_back(std::make_pair(
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.cc]: NOTE, not necessary every argument is memory pointer. We could pass in other integer variables(integer types) as uniform

File: src/runtime/opengl/opengl_module.cc [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        "arg" + std::to_string(i),
+        *static_cast<gl::Texture**>(void_args[i])
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.cc]: The order of the function arguments does not necessarily guarantee we write to the first argument, we can, however, detect this from the program and save this additional info as meta data

**Comment by phisiart** [comment]
Next step: fix things in OpenGL codegen.

**Comment by phisiart** [comment]
Update: I fixed the argument issue (I think?).

Now each OpenGL argument is categorized as one of the following three:
- Input texture;
- Output texture;
- Uniform value.

This information, together with argument names, need to be stored in some metadata. As a result, the "source" of an OpenGL module is now JSON.

I tested the changed version with `tests/webgl/test_local_save_load.py` and `tests/webgl/test_remote_save_load.py`. The latter is a simple ewise test using RPC with browser.

File: include/tvm/runtime/device_api.h (hunk 8 above)
Comment (phisiart) [reply include/tvm/runtime/device_api.h]: Changed.

File: include/tvm/schedule.h (hunk 11 above)
Comment (phisiart) [reply include/tvm/schedule.h]: Changed.

File: src/runtime/cpu_device_api.cc (hunk 9 above)
Comment (phisiart) [reply src/runtime/cpu_device_api.cc]: Changed.

File: src/runtime/rpc/rpc_session.h (hunk 10 above)
Comment (phisiart) [reply src/runtime/rpc/rpc_session.h]: Deleted.

File: src/runtime/opengl/opengl_module.cc (hunk 12 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.cc]: Changed.

File: src/runtime/opengl/opengl_module.cc (hunk 13 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.cc]: Changed.

File: Makefile (hunk 7 above)
Comment (phisiart) [reply Makefile]: Changed back.

**Comment by phisiart** [comment]
The next 2 correctness fixes to work on:

- Reduction axis not yet supported. This must be supported for gemm.
- ThreadIdx extent not utilized. This might lead to writing to out-of-bound elements in an array.

File: src/runtime/opengl/opengl_module.h [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+inline void OpenGLShader::Load(dmlc::JSONReader* reader) {
+  dmlc::JSONObjectReadHelper helper;
+  helper.DeclareField("arg_names", &arg_names);
+  helper.DeclareField("arg_kinds", &arg_kinds);
+  helper.DeclareField("source", &source);
+  helper.ReadAllFields(reader);
+}
+
+inline std::string ToJson(
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.h]: Json->JSON

File: src/runtime/opengl/opengl_module.h [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+namespace tvm {
+namespace runtime {
+
+enum OpenGLArgKind : int {
+  kInputTexture = 0,
+  kOutputTexture = 1,
+  kUniform = 2,
+};
+
+struct OpenGLShader {
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.h]: add a comment about this data structure

File: src/runtime/opengl/opengl_module.h [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  std::string source;
+  std::vector<std::string> arg_names;
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.h]: document the arg_kinds, refer to OpenGLArgKind

File: src/runtime/opengl/opengl_module.cc [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+  LOG(INFO) << "OpenGLWrappedFunc::operator()";
+
+  auto &shader = m_->GetShader(func_name_);
+  auto &program = m_->GetProgram(func_name_);
+  auto &func_info = m_->GetFunctionInfo(func_name_);
+
+  size_t nargs = shader.arg_kinds.size();
+
+  std::vector<std::tuple<std::string, TVMType, void*>> uniforms;
+  std::vector<std::pair<std::string, gl::Texture*>> inputs;
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.cc]: ```>> -> > >``` for some compilation issue with MSVC

File: src/runtime/opengl/opengl_module.cc [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+        output = *static_cast<gl::Texture**>(void_args[i]);
+        break;
+      }
+      default: {
+        LOG(FATAL) << "Invalid OpenGLArgKind";
+      }
+    }
+  }
+
+  m_->workspace().Render(program, uniforms, inputs, output);
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.cc]: I remember gl have API that is like setArg, which allows you to direct set Uniform and input of each object. Let us consider use that one as it might be more efficient than allocating a vector for each of these

File: tests/webgl/test_local_gemm.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
+from __future__ import absolute_import, print_function
+
```
Comment (tqchen) [inline tests/webgl/test_local_gemm.py]: write local test cases in normal testcase format so we can always use nose

File: tests/webgl/test_remote_save_load.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+"""
+
+import numpy as np
+import tvm
+from tvm.contrib import rpc, util, emscripten
+
+proxy_host = "localhost"
+proxy_port = 9090
+
+def test_remote_save_load():
```
Comment (tqchen) [inline tests/webgl/test_remote_save_load.py]: Make testcase compatible with nose-test. Change name of non-nose testable function to be other-name, like try_remote_save_load.

Always detect main when run the test function, so it won't get invoked by nose
 a flag "__main__", and

**Comment by phisiart** [comment]
Starting a massive cleanup.

File: src/codegen/codegen_opengl.cc [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+  LOG(INFO) << "CodeGenOpenGL::AddFunction";
+
+  // clear previous generated state.
+  this->InitFuncState(f);
+
+  this->decl_stream << "#version 300 es\n";
+  this->decl_stream << "precision highp float;\n";
+
+  this->decl_stream << "float tvm_get_texel(sampler2D tex, int idx) {\n";
+  this->decl_stream << "  return texelFetch(tex, ivec2(idx, 0), 0).r;\n";
```
[the remark this answers is not in the document]
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: I'm actually already using GL_R32F. It's in `TextureFormat GetTextureFormat(TVMType type)` in `opengl_device_api.cc`.

Sadly `texelFetch` returns a `vec4` anyway.

[the remark this answers is not in the document]
Comment (tqchen) [reply src/codegen/codegen_opengl.cc]: what is the implication? Does this mean we store data in compact format but the texFetch API will return a duplicated four floats?

[the remark this answers is not in the document]
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: I think `texelFetch` returns `vec4(the_real_stored_data, 0, 0, 0)`.

Notice that in deeplearnjs, they do something similar:
https://github.com/PAIR-code/deeplearnjs/blob/04fb4353854e6ab9a8f81b25036a819fa675d1ab/src/math/backends/webgl/render_ndarray_gpu_util.ts#L40

They use `texture2D()`, which takes float index instead of int index.

[the remark this answers is not in the document]
Comment (tqchen) [reply src/codegen/codegen_opengl.cc]: gotcha

**Comment by tqchen** [comment]
please let me know when changes are ready for review

**Comment by phisiart** [comment]
@tqchen I have a lot of `LOG(INFO)` for debugging. For a merge-able version, do I remove them or change them to `LOG(DEBUG)` or do something else?

**Comment by tqchen** [comment]
We do need to remove the LOG(INFO)s

File: tests/webgl/test_remote_save_load.py (hunk 20 above)
Comment (phisiart) [reply tests/webgl/test_remote_save_load.py]: Changed.

File: tests/webgl/test_local_gemm.py (hunk 19 above)
Comment (phisiart) [reply tests/webgl/test_local_gemm.py]: Changed.

File: src/runtime/opengl/opengl_module.h (hunk 14 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.h]: Changed.

File: src/runtime/opengl/opengl_module.h (hunk 15 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.h]: Added.

File: src/runtime/opengl/opengl_module.h (hunk 16 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.h]: I did something different: when outputing to JSON, transform to string. This gives better readability.

File: src/runtime/opengl/opengl_module.cc (hunk 17 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.cc]: Changed.

File: src/runtime/opengl/opengl_module.cc (hunk 18 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.cc]: I'm a little bit reluctant to do that. The reason is that all `gl*` API's as well as certain OpenGL concepts are encapsulated within `OpenGLWorkspace`.

**Comment by phisiart** [comment]
I think this version is ready for review.

File: src/codegen/codegen_opengl.cc [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+    os << "tvm_get_texel(" << GetVarID(buffer) << ", ";
+    PrintExpr(index, os);
+    os << ")";
+    return os.str();
+  }
+}
+
+void CodeGenOpenGL::PrintType(Type t, std::ostream& os) {
+  switch (t.code()) {
+    case halideir_type_int:
```
Comment (tqchen) [inline src/codegen/codegen_opengl.cc]: use kInt

File: src/codegen/codegen_opengl.cc [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+      CHECK_EQ(t.bits(), 32) << "Only support 32-bit int.";
+      os << "int";
+      break;
+    case halideir_type_uint:
```
Comment (tqchen) [inline src/codegen/codegen_opengl.cc]: kUInt

File: src/codegen/codegen_opengl.cc [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+      CHECK_EQ(t.bits(), 32) << "Only support 32-bit uint.";
+      os << "uint";
+      break;
+    case halideir_type_float:
```
Comment (tqchen) [inline src/codegen/codegen_opengl.cc]: kFloat

File: src/codegen/codegen_opengl.cc (hunk 22 above)
Comment (tqchen) [reply src/codegen/codegen_opengl.cc]: CHECK_EQ(t.lanes(), 1), as we don't support vector type yet?

File: src/codegen/codegen_opengl.h [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+  // Codegen for immediate values
+  void VisitExpr_(const IntImm* op, std::ostream& os) final;  // NOLINT(*)
+  void VisitExpr_(const UIntImm* op, std::ostream& os) final;  // NOLINT(*)
+  void VisitExpr_(const FloatImm* op, std::ostream& os) final;  // NOLINT(*)
+  void VisitExpr_(const StringImm* op, std::ostream& os) final;  // NOLINT(*)
+
+ private:
+  const Variable* output_;
+  std::unordered_set<const Variable*> inputs_;
+  const Variable* iter_var_;
```
Comment (tqchen) [inline src/codegen/codegen_opengl.h]: output_iter_var{nullptr}; // put initializer here

File: src/codegen/codegen_opengl.cc (hunk 23 above)
Comment (tqchen) [reply src/codegen/codegen_opengl.cc]: kDLUInt

File: src/codegen/codegen_opengl.cc [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+    // This is the output texture.
+    CHECK_EQ(index.get(), iter_var_)
+      << "GLSL must access corresponding elem of output texture.";
+    return GetVarID(buffer);
+  } else {
+    // This is an input texture.
+    this->inputs_.insert(buffer);
+
+    std::ostringstream os;
+    os << "tvm_get_texel(" << GetVarID(buffer) << ", ";
```
Comment (tqchen) [inline src/codegen/codegen_opengl.cc]: ```tvm_get_texel``` should be a member function of CodeGenOpenGL that returns the resulting string, instead of calling into function. This will allow us to add more overload in code generation mode.  Can we add support of 2D load for large tensors by doing textfetch( vec2(idx % const, idx /const)) where const equals 1024?

File: src/runtime/c_runtime_api.cc [hunk 27]
```diff
@@ -95,7 +96,8 @@ DeviceAPI* DeviceAPI::Get(TVMContext ctx, bool allow_missing) {
 }
 
 void* DeviceAPI::AllocWorkspace(TVMContext ctx, size_t size) {
-  return AllocDataSpace(ctx, size, kTempAllocaAlignment);
+  TVMType type_hint = {.code = kDLUInt, .bits = 8, .lanes = 1};
```
Comment (tqchen) [inline src/runtime/c_runtime_api.cc]: designated initializer is supported in C99 but not c++11 (need c++20), let us not do this for now

File: src/runtime/opengl/opengl_common.h [hunk 28]
```diff
@@ -0,0 +1,465 @@
...
+  Texture CreateTexture(TVMType type, size_t nbytes);
+
+  /*!
+   * \brief Upload user data into a sub-region of an OpenGL texture.
+   * \param texture The texture to be written to.
+   * \param begin The index of the first element to be written to.
+   * \param nelems The number of elements to be written to.
+   * \param data The user data.
+   */
+  void PutTextureData(Texture* texture, GLint begin, GLsizei nelems,
```
Comment (tqchen) [inline src/runtime/opengl/opengl_common.h]: style break multiple function argument into one argument per line.

File: src/runtime/opengl/opengl_module.cc [hunk 29]
```diff
@@ -0,0 +1,278 @@
...
+      }
+    }
+  }
+
+  ThreadWorkLoad wl = thread_axis_cfg_.Extract(args);
+  std::unique_ptr<GLint> thread_extent(new GLint(wl.block_dim(0)));
+  uniforms.emplace_back(shader.thread_extent_var, TVMType{kDLInt, 32, 1},
+                        static_cast<void*>(thread_extent.get()));
+
+  m_->workspace().Render(program, uniforms, inputs, output);
```
Comment (tqchen) [inline src/runtime/opengl/opengl_module.cc]: render only get called at one place, bring that logic here so that we don;t need to store intermediate result into vector of uniforms and inputs

File: src/schedule/schedule_lang.cc [hunk 30]
```diff
@@ -397,6 +397,46 @@ Stage& Stage::double_buffer() {
...
 
+Stage& Stage::opengl() {
+  CHECK(!is_scheduled()) << "Must be a fresh schedule";
+  StageNode *self = operator->();
+
+  auto all_iter_vars = self->all_iter_vars;  // curr version of all_iter_vars
+  CHECK(!all_iter_vars.empty()) << "At least one iter var";
+
+  // Fuse all dimensions to 1.
+  // TODO(zhixunt): Don't fuse reduction.
```
Comment (tqchen) [inline src/schedule/schedule_lang.cc]: remove todo

File: tests/webgl/test_remote_save_load.py [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+    fhost.import_module(fdev)
+
+    print('running...')
+    a = tvm.nd.array(np.random.uniform(size=16).astype(A.dtype), ctx)
+    b = tvm.nd.array(np.zeros(16, dtype=A.dtype), ctx)
+    c = tvm.nd.array(np.zeros(16, dtype=C.dtype), ctx)
+    fhost(a, b, c)
+    np.testing.assert_allclose(c.asnumpy(), a.asnumpy() + b.asnumpy())
+
+try_remote_save_load()
```
Comment (tqchen) [inline tests/webgl/test_remote_save_load.py]: chek main otherwise it get invoked

Comment (tqchen) [reply tests/webgl/test_remote_save_load.py]:
```python
if __name__ == "__main__":
    try_remote_save_load()
```

File: src/codegen/codegen_opengl.cc (hunk 26 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: Added member function `TexelFetch(const Variable* buffer, Expr index)`.

Yes, that's how to fetch texels when we support large tensors.

File: src/codegen/codegen_opengl.cc (hunk 22 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: Changed to kDLInt.
Added CHECK_EQ.

File: src/codegen/codegen_opengl.cc (hunk 23 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: Changed to kDLUInt.

File: src/codegen/codegen_opengl.cc (hunk 24 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.cc]: Changed to kDLFloat.

File: src/runtime/opengl/opengl_module.cc (hunk 29 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_module.cc]: Moved uniform & texture setting out of Render().

File: src/codegen/codegen_opengl.h (hunk 25 above)
Comment (phisiart) [reply src/codegen/codegen_opengl.h]: Changed `iter_var_` to `output_iter_var_` and added initializer.

File: src/runtime/c_runtime_api.cc (hunk 27 above)
Comment (phisiart) [reply src/runtime/c_runtime_api.cc]: Changed to `TVMType type_hint{kDLUInt, 8, 1}`.

File: src/runtime/opengl/opengl_common.h (hunk 28 above)
Comment (phisiart) [reply src/runtime/opengl/opengl_common.h]: Changed.

File: src/schedule/schedule_lang.cc (hunk 30 above)
Comment (phisiart) [reply src/schedule/schedule_lang.cc]: Removed.

File: tests/webgl/test_remote_save_load.py (hunk 31 above)
Comment (phisiart) [reply tests/webgl/test_remote_save_load.py]: Changed.

**Comment by phisiart** [comment]
Review comments are addressed.

**Comment by tqchen** [comment]
Thanks, this concludes the basic support of GL, and this PR is merged. Let us start new PRs to make further improvements
